tests/virt_autotest/virt_utils.pm

```diff
@@ -545,4 +545,23 @@ sub generate_testcase_xml {
     $xml_writer->endTag('testcase');
 }
 
+#Get installed SLE release version and service spack
+#get_sles_release only accepts /etc/os-release as source file
+#If sles version on vm guest needs to be parsed, vm guest name or ip address should be passed in
+#Of course sles version of other physical hosts can also be parsed by specifying its name or ip
+#For usage on locahost, no parameters need to be given
+sub get_sles_release {
+    my ($target_machine) = @_;
+    $target_machine //= '';
+    my $go_to_target      = $target_machine eq '' ? '' : "ssh root\@${target_machine}";
+    my $sles_release      = script_output("${go_to_target} cat /etc/os-release | grep -i version= | grep -iEo \"[0-9]{1,}(\\\.|\\\-)?(sp)?([0-9]{1,})?\"");
```

Comment:
```suggestion
    my ($target_command) = shift // 'cat /etc/os-release';
    my $sles_release      = script_output("${target_command} | grep -i version= | grep -iEo \"[0-9]{1,}(\\\.|\\\-)?(sp)?([0-9]{1,})?\"");
```

Reply:
It is no reasonable to rely on end user to pass in all things including "ssh root", which should be handled inside subroutine. The user only needs to specify the guest or host fqdn or ip address. But I will try to make these lines more compact.

Reply:
I think it is reasonable. My solution is way more universal and way more compact.

Reply:
Please have a look at my latest implementation. And the end user should only care about machine name or ip address, no other things.

Reply:
If you only care about localhost, then no parameter is needed, as there is default option:
```perl
my ($target_command) = shift // 'cat /etc/os-release';
```
With this approach you can also check other protocols than just SSH and other file locations when the filesystem is for example mounted somewhere else.

And it also saves two lines and one variable.

Reply:
In your implementation, target_command will never be empty. So if will be troublesome to handle whether "ssh root" should be prepend to it, let alone other protocol. So you have to ask user to use it like get_sles_release("ssh root\@aaabbbcc") which is not desirable.

Reply:
* I think `get_sles_release("ssh root\@AAABBBCC cat /etc/os-release")` is OK.
* Also `get_sles_release("cat /mnt/etc/os-release")` is possible.
* And be careful because `get_sles_release()` also works.

Reply:
Probably you should use get_sles_release("ssh root\\\@AAABBBCC cat /etc/os-release") to make it work eventually. I think this usage just complicates things. My original intention is to detect sles version on any running machine instead of os-release file anywhere. If you just want to parse a os-release file in someplace, then you want to achieve a different thing rather than perform testing on running SLES. For running os, ssh connection is way enough.

Reply:
I usually use `assert_script_run("ssh root\@$guest")` so `get_sles_release("ssh root\@AAABBBCC cat /etc/os-release")` should work.

As @okurz pointed out - there are already bunch of duplications where we are parsing the `os-release` file so instead of creating another one for **your very specific** use case I am trying to create **as universal function as possible** so we can continue to refactor our code by **reducing the number** of such duplications.

Reply:
I am also trying to do so, otherwise I just pick up scirpt_out and dump everything to its arguments. Now the subroutine wraps up very universal regex which can parse all sles release and returns both version and service pack info. I just do not want to dump everything to subroutine arguments which is not best practice for a subroutine. I would like to expand its functionality inside ranther than one all-round argument. At the same time, I do not object to an univeral function. My original thought is any future expansion to it is welcome and anyone can modify it just like we do and collaborate here in os-autoinst-distri-opensuse repo, so for this time the current implementation is fairly enough. But I think there is also a better solution here, I propose a better version of this subroutine like this:
get_sles_release($go_to_method, $release_file_path)
It can satisfy your need and, at the same time, looks nicer and much more resemble a function. $go_to_method can be "ssh root@macine" or somethig else, $release_file_path is so obvious(do not include "cat"). $Agree ?

Reply:
Agree. Can you then modify the code accordingly?

Reply:
Already changed

Reply:
OK, thank you Wei 👍